core/domain/subscription_services.py

```diff
@@ -268,6 +281,8 @@ def get_collection_ids_subscribed_to(user_id):
     """
     subscriptions_model = user_models.UserSubscriptionsModel.get(
         user_id, strict=False)
-    return (
-        subscriptions_model.collection_ids
-        if subscriptions_model else [])
+    if subscriptions_model:
+        collection_ids: List[str] = subscriptions_model.collection_ids
+        return collection_ids
```

Comment:
Here, `collection_ids` is a `datastore_services.StringProperty` and the return type of these properties is `Any`. So, to narrow down the type form `Any` to `List[str]`, I explicitly annotate them as List[str].

Reply:
Similar goes for `subscriptions_model.exploration_ids` because `exploration_ids` is also an instance of  `datastore_services.StringProperty` and the return type of this is also `Any`.

Reply:
Hmm, weird I thought that the types of models work correctly and that we don't need this kind of stuff.

Reply:
Ohh really sorry, I forgot to reply here!

In our stubs, we have defined the return value of the datastore's property as a Any type: 
https://github.com/oppia/oppia/blob/b5d0651e10bc75d63f121e0fc23b4fe364a1e053/stubs/google/cloud/ndb/__init__.pyi#L77
I think this is done on purpose because:
1. In the jobs folder `datastore's property` is considered as an instance of  `datastore_services.Property`.
2. In codebase, there are some places where a particular `datastore_services.Property`'s return value is considered as two different types. Like here: `StringProperty` is considered as str and List[str] as well.

I also tried to do `@ overload` here. but due to a very limited number of argument types, I'm not able to produce different function definitions which satisfy the `@ overload`'s requirement. 
 
If we use Union as a return type then we have to put assert in places where these properties are used.

Reply:
Ohh really sorry, I forgot to reply here!

In our stubs, we have defined the return value of the datastore's property as a Any type: 
https://github.com/oppia/oppia/blob/b5d0651e10bc75d63f121e0fc23b4fe364a1e053/stubs/google/cloud/ndb/__init__.pyi#L77
I think this is done on purpose because:
1. In the jobs folder `datastore's property` is considered as an instance of  `datastore_services.Property`.
2. In codebase, there are some places where a particular `datastore_services.Property`'s return value is considered as two different types. Like here: `StringProperty` is considered as str and List[str] as well.

I also tried to do `@ overload` here. but due to a very limited number of argument types, I'm not able to produce different function definitions which satisfy the `@ overload`'s requirement.

Reply:
@sahiljoster32 can you provide examples of where this happens?

```
In codebase, there are some places where a particular datastore_services.Property's return value is considered as two different types.
```

Also I think overloading makes sense but I don't understand your argument in (2). Could you elaborate on this a bit more?

We should try not to use Any here I think, it will have downstream effects on a bunch of things.

Reply:
Taking an example of `datastore_services.StringProperty` which is inherited from `datastore_services.Property`.

Eg 1: Here, `tags` is defined with `datastore_services.StringProperty` taking arguments `repeated=True, indexed=True` 
https://github.com/oppia/oppia/blob/b5d0651e10bc75d63f121e0fc23b4fe364a1e053/core/storage/collection/gae_models.py#L177
and it is considered as List[str] in https://github.com/oppia/oppia/blob/b5d0651e10bc75d63f121e0fc23b4fe364a1e053/core/domain/collection_services.py#L815.
 But on other hand, `owner_ids` is defined with same `datastore_services.StringProperty` taking same arguments `repeated=True, indexed=True`
https://github.com/oppia/oppia/blob/b5d0651e10bc75d63f121e0fc23b4fe364a1e053/core/storage/collection/gae_models.py#L816
and it is considered as `str` here 
https://github.com/oppia/oppia/blob/b5d0651e10bc75d63f121e0fc23b4fe364a1e053/core/storage/collection/gae_models.py#L897

Eg 2: Here, `ModelProperty` assumes that every `datastore_services.StringProperty` is an instance of `datastore_services.Property`.
https://github.com/oppia/oppia/blob/b5d0651e10bc75d63f121e0fc23b4fe364a1e053/core/jobs/types/model_property.py#L53
So, if we somehow change the return type of `datastore_services.StringProperty` to match it with str and List[str]. Then here in jobs file that return type fails because after changing the return type it is considered as an instance of str or list.

Why Overload is not successful here?
Overload is not successful here because overload requires different combinations of arguments to produce different return type. And here we have only 2 different arguments with the same type in each function definition.

```
class StringProperty(TextProperty):
    @overload
    def __get__(self, entity: Optional[Model], unused_cls: type[Model] | None = ..., indexed: bool=True, repeated: bool = True) -> List[str]: ... ( here we expect it to return both List[str] and str )

    @overload
    def __get__(self, entity: Optional[Model], unused_cls: type[Model] | None = ..., indexed: bool=True) -> str: ... ( This definition and above difnitation both fails for jobs files. )

    def __get__(self, entity: Optional[Model], unused_cls: type[Model] | None = ...) -> Any: ...

```

Reply:
Also, The other problem with overloads, in this case is that in some cases we are also providing other keyword arguments like : `Default =` and `choices =`. For this, we have to mention `**kwargs: Any` in overload. So if we define overloads as follows:
```
class StringProperty(TextProperty):
    @overload
    def __get__(self, entity: Optional[Model], unused_cls: type[Model] | None = ..., indexed: bool=True, repeated: bool = True, **kwargs: Any) -> List[str]: ... 

    @overload
    def __get__(self, entity: Optional[Model], unused_cls: type[Model] | None = ..., indexed: bool=True, **kwargs: ANy) -> str: ... 

    def __get__(self, entity: Optional[Model], unused_cls: type[Model] | None = ...) -> Any: ...
```

Then overload definition 2 is never matched because every keyword argument is covered inside definition 1 ( by **kwargs: Any)

Reply:
@seanlip and @vojtechjelinek PTAL!

Reply:
Hmm, I think `cls.owner_ids == user_id` is very specific case, as internally NDB evaluates this as `user_id in cls.owner_ids`.

Reply:
And we can maybe enforce the order of the kwargs so that it is always matched by MyPy overloads.

Reply:
```

> Hmm, I think `cls.owner_ids == user_id` is very specific case, as internally NDB evaluates this as `user_id in cls.owner_ids`. This should be possible to be ignored.

there are 80+ errors in this case!

Reply:
Also, I tried my best to cover these cases with overloads but I found that 2nd overload is never checked ( here I'm confused why this overload is not called, it would be helpful if you take a look )

```
class StringProperty(TextProperty):
    @overload
    def __get__(self, entity: Optional[Model], unused_cls: Type[Model] | None = ..., indexed: bool= ...) -> str: ... 
    
    @overload
    def __get__(self, entity: Optional[Model], unused_cls: Type[Model] | None = ..., indexed: Optional[bool] = ..., repeated: Optional[bool] = ...) -> List[str]: ... 
```
This is the overload that i used, Also i checked the Typeshed but i found that there also we are not using overloads instead we are using Union there. 
![Screenshot from 2022-06-23 15-40-18](https://user-images.githubusercontent.com/12553576/175275720-9d8f72c8-4b2b-4b62-9dd3-9c8a4ed2db7d.png)

Reply:
@vojtechjelinek PTAL!

Reply:
Hi @seanlip and @vojtechjelinek ,
I  have created an [issue](https://github.com/oppia/oppia/issues/15621) and added its corresponding TODO wherever it is applicable. Also, created a debugging doc which you may find [HERE](https://docs.google.com/document/d/1Qtc2pTv6Wcavt7CZcG5qVpIDAL1kye0-Vhp_yhT6-t0/edit?usp=sharing).

Reply:
Thanks @sahiljoster32! Sounds good -- I'm happy to defer to @vojtechjelinek for the rest of this. 

(Just for context, I only made the comment I did because I saw the note floating past my inbox and thought it wasn't really clear in explaining what was going on, so I figured I'd ask for clarification to help speed up the review process a bit.)

Reply:
Thanks for your comment!!, because previously I'm assuming it was a pattern that oppia follows for the ndb property but now I'm clear that it is a bug and we have to fix this ASAP. :)